### PR TITLE
mint circuit

### PR DESCRIPTION
**Pull request by 0xbok**

### CODE REVIEW

File: circuits/circuits/mint.circom [hunk 1]
```diff
@@ -1,20 +1,63 @@
...
+    signal initial_root_calculated <== CheckMerkleProof(levels)(
+        leaf <== 0,
+        pathElements <== pathElements,
+        pathIndices <== pathIndices
+    );
+    initial_root_calculated === initial_root;
+    new_root <== CheckMerkleProof(levels)(
+        leaf <== Poseidon(3)(inputs <== [recipient, leaf_coins[0], leaf_coins[1]]),
+        pathElements <== pathElements,
+        pathIndices <== pathIndices
```
Comment (BlakeMScurr) [inline circuits/circuits/mint.circom:58]: We can probably make this more succinct by using `UpdateLeaf`:
```
new_root <== UpdateLeaf(levels) (
    old_leaf <== 0,
    new_leaf <== Poseidon(3)(inputs <== [recipient, leaf_coins[0], leaf_coins[1]]),
    pathElements <== pathElements,
    pathIndices <== pathIndices,
    old_root <== initial_root
)
```

In fact, that applies to the mint tree as well. And *maybe* with the withdraw circuit too.

File: circuits/circuits/mint.circom [hunk 2]
```diff
@@ -1,20 +1,63 @@
...
 
 // If the sender is the 0 address, the transaction is a mint request,
 // if the recipient is 0, the transaction is a withdrawal, if neither are
 // 0 it's an L2->L2 send, and both can't be 0.
-template Mint(levels) {
+template Mint(levels, mintLevels) {
     signal input sender;
     signal input recipient;
+
+    signal input mint_root;
```
Comment (BlakeMScurr) [inline circuits/circuits/mint.circom:13]: This is good for now, but in the future we might be able to pass the mint request directly into circuit as an input. The main reason to do this is the gas cost of adding requests into the mint merkle tree onchain.

**Review by BlakeMScurr — APPROVED** [review]
LGTM, with comments

File: circuits/circuits/mint.circom (hunk 1 above)
Comment (0xbok) [reply circuits/circuits/mint.circom:58]: right, can make this change while doing the send circuit.

File: circuits/circuits/mint.circom (hunk 2 above)
Comment (0xbok) [reply circuits/circuits/mint.circom:13]: sounds good. we can add a _build mint tree_ circuit later.

**Pull request merged**